ezyrb/regular_grid.py

```diff
@@ -0,0 +1,131 @@
+    Multidimensional interpolator on regular grids.
+    See scipy's RegularGridInterpolator
+
+    """
+
+    def __init__(self):
+        self.interpolator = None
+        self.dim = None
+        self.n_modes = 0
+        self.mode_nr = None
+
+    def get_grid_axes(self, pts_scrmbld, vals_scrmbld):
+        """
+        calculates the grid axes from a meshed grid. The grid axes are given as
+        a tuple of ndarray of float, with shapes (m1, ), …, (mn, )
+        The values are ordered so they fit on a mesh generated with
+        numpy.meshgrid(ax1, ax2, ..., axn, indexing="ij")
+
+        """
+        # be aware of floating point precision in points!
+        grid_axes = []
+        iN = []  # index in dimension N
+        nN = []  # size of dimension N
+        dim = pts_scrmbld.shape[1]
+        for i in range(dim):
+            xn, unique_inverse_n = np.unique(pts_scrmbld[:, i],
+                                             return_inverse=True)
+            grid_axes.append(xn)
+            nN.append(len(xn))
+            iN.append(unique_inverse_n)
+
+        if np.prod(nN) != len(vals_scrmbld):
+            raise ValueError("points and values are not on a regular grid")
+        new_row_index = calculate_flat_index(iN, nN)
+        reverse_scrambling = np.argsort(new_row_index)
+        vals_on_regular_grid = vals_scrmbld[reverse_scrambling, :]
+        return grid_axes, vals_on_regular_grid
+
+    def fit(self, points, values, **kvargs):
+        """
+        Construct the interpolator given `points` and `values`.
+        Assumes that the points are on a regular grid, fails when not.
+        see scipy.interpolate.RegularGridInterpolator
+
+        :param array_like points: the coordinates of the points.
+        :param array_like values: the values in the points.
+        """
+        # we have two options
+        # 1.: we could make an interpolator for every mode and its coefficients
+        # or 2.: we "interpolate" the mode number
+        # option 1 is cleaner, but option 2 performs better
+        # X = U S VT, X being shaped (m, n)
+        points = np.array(points)
+        if not np.issubdtype(points.dtype, np.number):
+            raise ValueError('Invalid format or dimension for the argument'
+                             '`points`.')
+
+        if len(points.shape) == 1:
+            points.shape = (-1, 1)
+
+        self.dim = len(points[0])
+        vals = np.asarray(values)
+        grid_axes, values_grd = self.get_grid_axes(points, vals)
+        self.n_modes = vals.T.shape[0]  # vals = (S@VT).T = S@V
+        if self.n_modes > 1:
+            self.mode_nr = np.arange(self.n_modes)
+            extended_grid = [self.mode_nr, *grid_axes]
+            shape = [self.n_modes, ]
+        else:
+            extended_grid = grid_axes
+            shape = []
+        for i in range(self.dim):
+            shape.append(len(grid_axes[i]))
+        assert np.prod(shape) == values_grd.size, "Values don't match grid. "\
```

Comment:
Add a ValueError

Reply:
I moved it entirely, I thought here is not the right place since get_grid_axes() has to ensure this